File: Phase3/src/item.cpp
#include "../include/item.h"

Item::Item(const std::string& itemName, User* seller, User* topBidder, const int& auctionLengthRemaining, const float& currentHighestBid) {
    m_itemName = itemName;
    m_seller = seller;
    m_topBidder = topBidder;
    m_auctionLengthRemaining = auctionLengthRemaining;
    m_currentHighestBid = currentHighestBid;
}

Item::~Item(){
    delete m_seller;
    delete m_topBidder;
}

const std::string& Item::getItemName() const {
    return m_itemName;
}

User* Item::getSeller() const {
    return m_seller;
}

User* Item::getTopBidder() const {
    return m_topBidder;
}

const int& Item::getAuctionLengthRemaining() const {
    return m_auctionLengthRemaining;
}

const float& Item::getCurrentHighestBid() const {
    return m_currentHighestBid;
}

// Overrides to string method
std::ostream& operator<<(std::ostream& os, const Item& obj) {
    // Formats item name to 25 characters
    std::string itemName = obj.getItemName();
    itemName.resize(25, ' ');

    // Formats seller name to 15 characters
    std::string sellerName = obj.getSeller()->getUsername();
    sellerName.resize(15, ' ');

    // Formats top bidder name to 15 characters
    std::string topBidderName = obj.getTopBidder()->getUsername();
    topBidderName.resize(15, ' ');

    // Declares and initializes stringstream
    std::stringstream ss;

    // Formats auction length remaining to 3 characters
    int auctionLengthRemainngVal = (int)obj.getAuctionLengthRemaining();
    ss << std::setfill('0') << std::setw(3) << auctionLengthRemainngVal;
    std::string auctionLengthRemaining = ss.str();
    ss.str(std::string());  // Clears stringstream

    // Formats current highest bid to 6 characters
    int currentHighestBidVal = (int)obj.getCurrentHighestBid();
    ss << std::setfill('0') << std::setw(3) << currentHighestBidVal << ".00";
    std::string currentHighestBid = ss.str();
    ss.str(std::string());  // Clears stringstream
    
    return (os << itemName << " " << sellerName << " " << topBidderName << " " << auctionLengthRemaining << " " << currentHighestBid);
}

File: Phase4/backend/src/item.py
from Users.user import User

class Item:
    def __init__(self, item_name: str, seller: User, top_bidder: User, days_left: int, top_bid: float):
        self.item_name = item_name
        self.seller = seller
        self.top_bidder = top_bidder
        self.days_left = days_left
        self.top_bid = top_bid
        
    def __hash__(self) -> int:
        return hash(self.item_name)

    def __str__(self):
        return f"Item({self.item_name},{self.seller},{self.top_bidder},{self.days_left},{self.top_bid})"
    
    def __eq__(self, other):
        return self.item_name == other.item_name and self.seller == other.seller and self.top_bidder == other.top_bidder and self.days_left == other.days_left and self.top_bid == other.top_bid
    